src/Bridge.ts

```diff
@@ -0,0 +1,130 @@
+import Emittery from "emittery";
+import debugFactory from "debug";
+import _get from "lodash.get";
+import { v4 as uuid } from "uuid";
+import {
+  GET_REQUEST,
+  HANDHSAKE_START,
+  HANDSHAKE_REPLY,
+  PARENT_EMIT,
+  MESSAGE_TYPE,
+} from "./constants";
+import { createParentEmit } from "./events";
+
+import { getResponse, IGetRequest, IGetResponse, IParentEmit } from "./events";
+const debug = debugFactory("ibridge:parent");
+
+export interface INativeEventData {
+  type: typeof MESSAGE_TYPE;
+  eventName: string;
+  data: unknown;
+}
+
+export interface IConstructorArgs {
+  window: Window; // TODO others
+  // The window needs to be fully loaded
+  // i.e. the iframe has triggered the onLoad event
+  remoteWindow: Window; // TODO  others
+  remoteOrigin?: string;
+  model: object;
+}
+
+// TODO this needs to be generic on TModel
+// TODO support other types of windows like webworkers,
+// modal windows, etc
+export default class Bridge extends Emittery {
+  window: Window;
+  remoteWindow: Window;
+  remoteOrigin?: string;
+  model: object;
+  context: any = {};
+
+  constructor(args: IConstructorArgs) {
+    super();
+
+    this.window = args.window;
+    this.remoteWindow = args.remoteWindow;
+    this.remoteOrigin = args.remoteOrigin;
+    this.model = args.model;
+
+    // TODO does this work with webworkers? otherwise
+    // use addEventListener
+    this.remoteWindow.onmessage = this.dispatcher.bind(this);
+    this.on(GET_REQUEST, this.handleGet.bind(this) as any);
+  }
+
+  private dispatcher(event: MessageEvent): void {
+    // TODO validate event (origin? data shape?)
+    // TODO rework the events module to contain all the necesary
+    // event definition, types, validations, etc
+    //if (!isValidEvent(event, this.remoteOrigin)) {
+    //return;
+    //}
+
+    const { eventName, data } = event.data as INativeEventData;
+    this.emitToLocal(eventName, data);
+  }
+
+  // This is only to avoid confusions between Emittery's
+  // this.emit and this.emitToRemote, as a rule of thumb this.emit
+  // should only be used by the dispatcher, the restof the code 
+  // and even the lib consumers should use emitToRemote
+  private emitToLocal(eventName: string, data?: unknown): void {
+    super.emit(eventName, data)
+  }
+
+  async emit(eventName: string, data?: unknown): Promise<void> {
+    console.warn("this function is private and has now effect, you should probably want to use emitToRemote")
+  }
+
+  emitToRemote(eventName: string, data?: unknown): void {
+    debug(`emit "%s" with data %O`, eventName, data);
```

Comment:
since Bridge is a module that is going to be used from both "parent" and "child" perhaps we want to make it an instance variable with a variable prefix so that we get differentiated output when debugging